### ISSUE TITLE
build: decide on policy for supporting Go on Microsoft Linux 

### ISSUE THREAD

**Issue by arekkas** [golang/go#16628]
1. What version of Go are you using (`go version`)?

Go1.4 is installed from binaries: `go version go1.4 linux/amd64`

2. What operating system and processor architecture are you using (`go env`)?

I am trying out the new bash on ubuntu on windows 10 (amd64) feature: https://msdn.microsoft.com/en-us/commandline/wsl/about

```bash
$ go env
GOARCH="amd64"
GOBIN=""
GOCHAR="6"
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPATH="/mnt/d/workspace/go"
GORACE=""
GOROOT="/home/foobar/.gvm/gos/go1.4"
GOTOOLDIR="/home/foobar/.gvm/gos/go1.4/pkg/tool/linux_amd64"
CC="gcc"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0"
CXX="g++"
CGO_ENABLED="1"
```

3. What did you do?
If possible, provide a recipe for reproducing the error.
A complete runnable program is good.
A link on play.golang.org is best.

I tried installing go 1.6 from source using [gvm](https://github.com/moovweb/gvm):

```bash
# install go1.4 from binaries
gvm install go1.4 -B
gvm use go1.4
export GOROOT_BOOTSTRAP=$GOROOT

# install go1.6 from source
gvm install go1.6
Installing go1.6...
 * Compiling...
ERROR: Failed to compile. Check the logs at /home/foobar/.gvm/logs/go-go1.6-compile.log
ERROR: Failed to use installed version
```

4. What did you expect to see?

I expected Go to be compiled

5. What did you see instead?

```bash
$ cat  /home/foobar/.gvm/logs/go-go1.6-compile.log
##### Building Go bootstrap tool.
cmd/dist
# _/home/foobar/.gvm/gos/go1.6/src/cmd/dist
%L%: internal compiler error: fault
```

**Comment by arekkas** [comment]
The same thing happens when running `go run ...` on any go code:

```
$ go run main.go
# github.com/Sirupsen/logrus
%L%: internal compiler error: fault
# github.com/julienschmidt/httprouter
%L%: internal compiler error: fault
# github.com/ory-am/common/rand/sequence
%L%: internal compiler error: fault
# github.com/go-errors/errors
%L%: internal compiler error: fault
# github.com/asaskevich/govalidator
%L%: internal compiler error: fault
# github.com/pkg/errors
%L%: internal compiler error: fault
# golang.org/x/crypto/blowfish
%L%: internal compiler error: fault
....
```

**Comment by davecheney** [comment]
The windows 10 emulation is not perfect. I recommend you use a newer
version of go as your bootstrap compile.

**Comment by arekkas** [comment]
Thank you, that got me a step further. I am now using go1.6 from binaries to compile go1.7rc4, which still fails, unfortunately:

```bash
$ cat /home/foobar/.gvm/logs/go-go1.7rc4-compile.log
##### Building Go bootstrap tool.
cmd/dist

##### Building Go toolchain using /home/foobar/.gvm/gos/go1.6.
bootstrap/internal/sys
bootstrap/internal/obj
bootstrap/internal/obj/arm
bootstrap/internal/obj/arm64
bootstrap/internal/obj/mips
bootstrap/internal/obj/ppc64
bootstrap/internal/obj/s390x
bootstrap/internal/obj/x86
bootstrap/asm/internal/arch
bootstrap/asm/internal/flags
bootstrap/asm/internal/lex
bootstrap/asm/internal/asm
bootstrap/internal/bio
bootstrap/asm
bootstrap/compile/internal/big
bootstrap/compile/internal/ssa
bootstrap/internal/gcprog
bootstrap/compile/internal/gc
bootstrap/compile/internal/amd64
bootstrap/compile/internal/arm
bootstrap/compile/internal/arm64
bootstrap/compile/internal/mips64
bootstrap/compile/internal/ppc64
bootstrap/compile/internal/s390x
bootstrap/compile/internal/x86
bootstrap/compile
bootstrap/link/internal/ld
bootstrap/link/internal/amd64
bootstrap/link/internal/arm
bootstrap/link/internal/arm64
bootstrap/link/internal/mips64
bootstrap/link/internal/ppc64
bootstrap/link/internal/s390x
bootstrap/link/internal/x86
bootstrap/link

##### Building go_bootstrap for host, linux/amd64.
runtime/internal/sys
runtime/internal/atomic
go tool dist: FAILED: /home/foobar/.gvm/gos/go1.7rc4/pkg/tool/linux_amd64/asm -I /tmp/go-tool-dist-364143704/runtime/internal/atomic -I /home/foobar/.gvm/gos/go1.7rc4/pkg/include -D GOOS_linux -D GOARCH_amd64 -D GOOS_GOARCH_linux_amd64 -o /tmp/go-tool-dist-364143704/runtime/internal/atomic/asm_amd64.o /home/foobar/.gvm/gos/go1.7rc4/src/runtime/internal/atomic/asm_amd64.s: wait: interrupted system call
go tool dist: open /tmp/go-tool-dist-364143704/runtime/internal/atomic/_go_.a: no such file or directory
```

What's weird, running `gvm install go1.7rc4` consecutive times results in a variety of error messages, for example:

```
##### Building go_bootstrap for host, linux/amd64.
runtime/internal/sys
runtime/internal/atomic
runtime
go tool dist: FAILED: /home/foobar/.gvm/gos/go1.7rc4/pkg/tool/linux_amd64/asm -I /tmp/go-tool-dist-949347488/runtime -I /home/foobar/.gvm/gos/go1.7rc4/pkg/include -D GOOS_linux -D GOARCH_amd64 -D GOOS_GOARCH_linux_amd64 -o /tmp/go-tool-dist-949347488/runtime/asm_amd64.o /home/foobar/.gvm/gos/go1.7rc4/src/runtime/asm_amd64.s: read |0: interrupted system call
unicode/utf16
unicode/utf8
encoding
errors
internal/race
internal/syscall/windows/sysdll
math
sort
sync/atomic
unicode
go tool dist: FAILED: /home/foobar/.gvm/gos/go1.7rc4/pkg/tool/linux_amd64/compile -pack -o /tmp/go-tool-dist-949347488/sort/_go_.a -p sort /home/foobar/.gvm/gos/go1.7rc4/src/sort/search.go /home/foobar/.gvm/gos/go1.7rc4/src/sort/sort.go: wait: interrupted system call
go tool dist: FAILED: /home/foobar/.gvm/gos/go1.7rc4/pkg/tool/linux_amd64/compile -pack -o /tmp/go-tool-dist-949347488/internal/syscall/windows/sysdll/_go_.a -p internal/syscall/windows/sysdll /home/foobar/.gvm/gos/go1.7rc4/src/internal/syscall/windows/sysdll/sysdll.go: fork/exec /home/foobar/.gvm/gos/go1.7rc4/pkg/tool/linux_amd64/compile: interrupted system call
go tool dist: open /tmp/go-tool-dist-949347488/sync/atomic/asm_amd64.o: no such file or directory
can't create /tmp/go-tool-dist-949347488/sort/_go_.a: open /tmp/go-tool-dist-949347488/sort/_go_.a: no such file or directory
can't create /tmp/go-tool-dist-949347488/math/_go_.a: open /tmp/go-tool-dist-949347488/math/_go_.a: no such file or directory
can't create /tmp/go-tool-dist-949347488/unicode/_go_.a: open /tmp/go-tool-dist-949347488/unicode/_go_.a: no such file or directory
```

**Comment by davecheney** [comment]
Real talk: the Windows 10 Linux thing isn't complete, some stuff isn't working and that isn't Go's fault. I've heard that go 1.6.3 works well.

**Comment by arekkas** [comment]
That explanation works for me ;)

**Comment by bradfitz** [comment]
You can file bugs against Microsoft Linux at https://github.com/Microsoft/BashOnWindows/issues

**Comment by minux** [comment]
Please try Go 1.7.

**Comment by arekkas** [comment]
> Please try Go 1.7.

Compiling regular Go code with Go 1.6 works fine. Compiling Go 1.7 using Go 1.6 doesn't work. Compiling Go 1.7rc2 using Go 1.7rc1 (installed from binaries) doesn't work either.

**Comment by arekkas** [comment]
Compiling Go 1.7rc2 using Go 1.7rc1 results in:

```
##### Building Go bootstrap tool.
cmd/dist
go build _/home/foobar/.gvm/gos/go1.7rc2/src/cmd/dist: /home/foobar/.gvm/gos/go1.7rc1/pkg/tool/linux_amd64/compile: waitid: function not implemented
```

which looks to me like there's some APIs missing?

**Comment by minux** [comment]
Please wait for the next 1.7rc (1.7rc6) for the waitid fix.

**Comment by arekkas** [comment]
> Please wait for the next 1.7rc (1.7rc6) for the waitid fix.

I will! I'll also reopen this issue as it seems of interest

**Comment by minux** [comment]
As the linux subsystem of Windows 10 is fairly new,
I just set the milestone to Go 1.8. RC6 is due today,
and I doubt we are going to delay Go 1.7 release
any longer even if it doesn't work on Linux subsystem
of Windows 10.

**Comment by ianlancetaylor** [comment]
The `waitid` bug is #16610.

**Comment by jboelter** [comment]
This issue is tracking SA_RESTART support (https://github.com/Microsoft/BashOnWindows/issues/349)

You can upvote it here: https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo/suggestions/14046789-add-support-for-restartable-system-calls-i-e-sa

**Comment by arekkas** [comment]
> Support for restartable syscalls shipped in Windows 10 Insiders build 14905: https://msdn.microsoft.com/en-us/commandline/wsl/release_notes#build-14905

**Comment by quentinmit** [comment]
waitid is now implemented.

@arekkas Can you please try the latest version of Go 1.7 against the latest version of Windows?

@rsc I hear you were working on getting "bash on windows" running; if you have success, can you try running all.bash to see how close we are to passing?

**Comment by quentinmit** [comment]
As for policy: I think we should support Linux-on-Windows as a second-class port. We should run a builder for it, but not let broken builds block the release.

**Comment by rsc** [comment]
I failed at getting the "Anniversary Edition" of Windows onto my laptop despite repeated updates. It just didn't want to offer it to me, and you need that one to get to the Ubuntu developer mode. Even though I didn't get it installed, the docs I saw online had it clearly marked as very beta. It's OK if we want to put fixes in, but it's not going to block Go 1.8.

As far as "make a decision", it sounds like other people were saying the same thing, so I think we've decided.

**Comment by quentinmit** [comment]
FYI, #17365 is tracking our setup of a Linux-on-Windows builder. We'll try to get this done in time for Go 1.9, but given that Windows Server 2016 doesn't support Linux, I'm not sure how we'll get there.

There are still problems with all.bash on Linux on Windows 10, more details in that issue.
